File: app/src/main/java/de/christinecoenen/code/zapp/app/livestream/ui/list/ChannelListAdapter.java
package de.christinecoenen.code.zapp.app.livestream.ui.list;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.WeakHashMap;

import de.christinecoenen.code.zapp.app.livestream.ui.views.ProgramInfoViewBase;
import de.christinecoenen.code.zapp.databinding.FragmentChannelListItemBinding;
import de.christinecoenen.code.zapp.models.channels.ChannelModel;
import de.christinecoenen.code.zapp.models.channels.IChannelList;

class ChannelListAdapter extends RecyclerView.Adapter<ChannelListAdapter.ViewHolder> {

	private static final String TAG = ChannelListAdapter.class.getSimpleName();

	private final IChannelList channelList;
	private final LayoutInflater inflater;
	private final WeakHashMap<ViewHolder, ?> visibleViews = new WeakHashMap<>();
	private final Listener listener;

	ChannelListAdapter(Context context, IChannelList channelList, Listener listener) {
		this.channelList = channelList;
		this.listener = listener;

		inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
		setHasStableIds(true);
	}

	void pause() {
		for (ViewHolder holder : visibleViews.keySet()) {
			holder.pause();
		}
	}

	void resume() {
		notifyDataSetChanged();
		for (ViewHolder holder : visibleViews.keySet()) {
			holder.resume();
		}
	}

	@NonNull
	@Override
	public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
		FragmentChannelListItemBinding binding = FragmentChannelListItemBinding.inflate(inflater, parent, false);
		return new ViewHolder(binding);
	}

	@Override
	public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
		ChannelModel channel = channelList.get(position);
		holder.setChannel(channel);
	}

	@Override
	public void onViewDetachedFromWindow(@NonNull ViewHolder holder) {
		holder.pause();
		visibleViews.remove(holder);
	}

	@Override
	public void onViewAttachedToWindow(@NonNull ViewHolder holder) {
		holder.resume();
		visibleViews.put(holder, null);
	}

	@Override
	public int getItemCount() {
		return channelList.size();
	}

	@Override
	public long getItemId(int position) {
		return position;
	}

	class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener, View.OnLongClickListener {

		private TextView subtitle;
		private ImageView logo;
		private ProgramInfoViewBase programInfoView;

		private ChannelModel channel;

		ViewHolder(FragmentChannelListItemBinding binding) {
			super(binding.getRoot());

			binding.getRoot().setOnLongClickListener(this);
			binding.getRoot().setOnClickListener(this);

			subtitle = binding.textChannelSubtitle;
			logo = binding.imageChannelLogo;
			programInfoView = binding.programInfo;
		}

		@Override
		public void onClick(View view) {
			if (channel != null) {
				ChannelListAdapter.this.listener.onItemClick(channel);
			}
		}

		@Override
		public boolean onLongClick(View view) {
			if (channel != null) {
				ChannelListAdapter.this.listener.onItemLongClick(channel, view);
			}
			return true;
		}

		void setChannel(ChannelModel channel) {
			this.channel = channel;

			logo.setImageResource(channel.getDrawableId());
			logo.setContentDescription(channel.getName());

			if (channel.getSubtitle() == null) {
				subtitle.setVisibility(View.GONE);
			} else {
				subtitle.setText(channel.getSubtitle());
				subtitle.setVisibility(View.VISIBLE);
			}

			programInfoView.setChannel(channel);
		}

		void pause() {
			programInfoView.pause();
		}

		void resume() {
			programInfoView.resume();
		}
	}

	public interface Listener {
		void onItemClick(ChannelModel channel);

		void onItemLongClick(ChannelModel channel, View view);
	}
}

File: app/src/main/java/de/christinecoenen/code/zapp/app/livestream/ui/list/ChannelListAdapter.kt
package de.christinecoenen.code.zapp.app.livestream.ui.list

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import de.christinecoenen.code.zapp.R
import de.christinecoenen.code.zapp.app.livestream.ui.views.ProgramInfoViewBase
import de.christinecoenen.code.zapp.model.ChannelModel
import de.christinecoenen.code.zapp.model.IChannelList
import kotlinx.android.synthetic.main.fragment_channel_list_item.view.*
import java.util.*

internal class ChannelListAdapter(context: Context,
								  private val channelList: IChannelList,
								  private val listener: Listener) : RecyclerView.Adapter<ChannelListAdapter.ViewHolder>() {

	private val inflater: LayoutInflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
	private val visibleViews = WeakHashMap<ViewHolder, Any>()

	init {
		setHasStableIds(true)
	}

	fun pause() {
		visibleViews.keys.forEach { it.pause() }
	}

	fun resume() {
		notifyDataSetChanged()
		visibleViews.keys.forEach { it.resume() }
	}

	override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
		val view = inflater.inflate(R.layout.fragment_channel_list_item, parent, false)
		return ViewHolder(view)
	}

	override fun onBindViewHolder(holder: ViewHolder, position: Int) {
		val channel = channelList[position]
		holder.setChannel(channel)
	}

	override fun onViewDetachedFromWindow(holder: ViewHolder) {
		holder.pause()
		visibleViews.remove(holder)
	}

	override fun onViewAttachedToWindow(holder: ViewHolder) {
		holder.resume()
		visibleViews[holder] = null
	}

	override fun getItemCount(): Int {
		return channelList.size()
	}

	override fun getItemId(position: Int): Long {
		return position.toLong()
	}

	internal inner class ViewHolder(view: View) : RecyclerView.ViewHolder(view), View.OnClickListener, View.OnLongClickListener {

		var subtitle: TextView = view.text_channel_subtitle
		var logo: ImageView = view.image_channel_logo
		var programInfoView: ProgramInfoViewBase = view.program_info

		private lateinit var channel: ChannelModel

		init {
			view.setOnLongClickListener(this)
			view.setOnClickListener(this)
		}

		override fun onClick(view: View) {
			listener.onItemClick(channel)
		}

		override fun onLongClick(view: View): Boolean {
			listener.onItemLongClick(channel, view)
			return true
		}

		fun setChannel(channel: ChannelModel) {
			this.channel = channel

			logo.setImageResource(channel.drawableId)
			logo.contentDescription = channel.name

			if (channel.subtitle == null) {
				subtitle.visibility = View.GONE
			} else {
				subtitle.text = channel.subtitle
				subtitle.visibility = View.VISIBLE
			}

			programInfoView.setChannel(channel)
		}

		fun pause() {
			programInfoView.pause()
		}

		fun resume() {
			programInfoView.resume()
		}
	}

	interface Listener {
		fun onItemClick(channel: ChannelModel)

		fun onItemLongClick(channel: ChannelModel, view: View)
	}
}
